### PR TITLE
Different async action mapping approach

### LINKED ISSUE #3576: Events stopped being added for an Action, whilst still working for identical duplicate Action
## Bug description

We have an action defined for a custom event "user signed up" (ID: 32) that filters where a property "test_user = false" and "gender = female". Since one month there are no new events showing in the UI and none of our slack messages are triggered. Today we tried to make a new action (ID: 119) with the same configuration and the events then show correctly.

**DB Query to examine action configuration**
It appears the action match group steps are configured exactly the same:

```
SELECT action_id, event, properties from posthog_actionstep where action_id in (32, 119);

# Old Action (32)
32
"user signed up"
[{"key": "gender", "type": "person", "value": "female", "operator": "exact"}, {"key": "test_user", "type": "person", "value": "false", "operator": "exact"}]

# New Action (119)
119
"user signed up"
[{"key": "gender", "type": "person", "value": "female", "operator": "exact"}, {"key": "test_user", "type": "person", "value": "false", "operator": "exact"}]
```

**DB Query to examine matched events**

```
SELECT action_id, count(*) from posthog_action_events where action_id in (32, 119) GROUP by action_id;
32	2159
119	2383
```

## Expected behavio(u)r

We expect the action to perform reliably in capturing events for both display in dashboards (this is how we noticed - the numbers didn't match our internal database) as well as the slack notifications that we use.

## How to reproduce

Not sure how to get into this state as we can't see any clear difference and when we create a "new" action with the same configuration it seems to work fine. My gut feeling is there might be some cache issues but not sure.

## Environment

- [ ] PostHog self-hosted with Kubernetes 1.9, Helm chart 1.4.8, posthog version: 1.21.0

## Additional context

One month ago when the events stopped for this Action we were on Posthog 1.19.0 I think. We also noticed another Action with the same issue which we "fixed" by deleting and re-adding the filters but this is obviously not ideal. Hence with this action we decided not to modify it so that we can try and investigate what the root cause of the issue is.

### PR DESCRIPTION
**Pull request by Twixes**
## Changes

This makes async calculation of new action-event pairs every 30 seconds the new and default approach.

Benefits over the previous async approach:
- enabled by default on Postgres whether plugin server ingesting or not
- actually tested since it's the default
- 30 seconds instead of the previous 5 minutes should make the user experience understandable (only a slight delay), while still being reasonably efficient (only calculating for events since last calculation)

Benefits over the previous on-the-fly approach:
- most importantly compatible with plugins
- doesn't require the plugin server additionally hammering Redis/Celery with webhook task dispatches, since that's also fired every 30 s
- overall significantly less queries

In the future we may want to hand off action calculation to the plugin server (PostHog/plugin-server#235), but that will require a serious effort with exact porting of the matching logic, which we do not want to do now.

Closes #3627 (Marius's WIP PR). Resolves #3576.

## Checklist

- [x] All querysets/queries filter by Organization, by Team, and by User
- [x] Django backend tests

### CODE REVIEW

File: posthog/models/action.py [hunk 1]
```diff
@@ -36,49 +37,42 @@ def calculate_events(self, start=None, end=None):
...
-            except Exception as err:
-                capture_exception(err)
-
-        self.is_calculating = False
-        self.last_calculated_at = calculated_at
-        self.save()
+            delete_query = f"""DELETE FROM "posthog_action_events" WHERE "action_id" = {self.pk}"""
+
+            if not recalculate_all:
+                delete_query += f""" AND "event_id" IN (
```
Comment (macobo) [inline posthog/models/action.py:46]: My worry here is that this will fall over hard on any larger postgres instance.

What does the query plan look like for this subquery?

File: posthog/models/action.py [hunk 2]
```diff
@@ -36,49 +37,42 @@ def calculate_events(self, start=None, end=None):
...
+
+            cursor = connection.cursor()
+            with transaction.atomic():
+                try:
+                    cursor.execute(delete_query + ";" + insert_query, params)
+                except Exception as err:
+                    capture_exception(err)
+
+            if self.post_to_slack:
+                for event in self.events.filter(
```
Comment (macobo) [inline posthog/models/action.py:66]: This seems like it could lead to double-deliveries right?

1. time=0 calculate_events gets called, stalls before this (e.g. on the insert/delete)
2. time=30 calculate_events gets called again
3. Query from time=30 resolves, bunch of events sent to be posted for webhooks
4. Query from time=0 resolves, bunch of events get sent to be posted for webhooks.

File: posthog/models/action.py (hunk 1 above)
Comment (macobo) [reply posthog/models/action.py:46]: Thinking about it more - do we really need this delete query if we're not recalculating all? 

The below insert query should handle properly the case where we calculate the same event multiple times and if action changes that should trigger a full recalculation.

File: posthog/models/action.py (hunk 2 above)
Comment (macobo) [reply posthog/models/action.py:66]: Also when we Insert + Delete (above) and insert again

File: posthog/models/action.py (hunk 1 above)
Comment (Twixes) [reply posthog/models/action.py:46]: It should not fall over, in fact this has been our recommended approach for larges Postgres instances, since on-the-fly was hammering Redis and Postgres really hard. But TBH it's difficult to analyze a query plan here, as this is very different depending on what actions an instance has defined.

I kind of agree that this should be practically the same without the `DELETE`, but I was removed from the initial implementation of this periodic mapping approach so there may be some context I'm missing. 🤔

Comment (macobo) [reply posthog/models/action.py:46]: I'm not sure I understood the above. _what_ is the recommended approach for Postgres instances?

--- 

The original comments where specifically about doing the DELETE when not doing a full recalculate. This seems bad due to:
1. *Performance* (1) - Doing an event table query (without a team_id filter) is likely to be slow on larger instances. This could easily take minutes if the query planner decides to do a full table scan (hence the question on query plans)
2. *Performance* (2) - Deletes result in trashing in the table and extra vacuuming needs to be done on the postgres instance to clear things up.
3. *Correctness* - Given that the below INSERT statement ignores duplicate rows (via the ON CONFLICT clause), I don't think this is needed for the query to be correct. 

I'm not 100% sure on point (3) this though - is there a hidden reason why we need to delete data every 30s for this to work?

Comment (macobo) [reply posthog/models/action.py:45]: Looking at this one more time I see I missed things on my first read - this query was as such previously as well. I still wonder about (3) though - why is it needed?

File: posthog/models/action.py [hunk 3]
```diff
@@ -24,61 +24,54 @@ def calculate_events(self, start=None, end=None):
...
+            )
+
+            cursor = connection.cursor()
+            with transaction.atomic():
+                try:
+                    cursor.execute(delete_query + ";" + insert_query, params)
+                except Exception as err:
+                    capture_exception(err)
+
+            if self.post_to_slack:
```
Comment (macobo) [inline posthog/models/action.py:64]: I think we only do this for periodic recalculations rather than full recalculations.

Scenario: User changes definition for "user signed up". With this change I think the user would suddenly get 20k notifications in slack.

File: posthog/tasks/webhooks.py [hunk 4]
```diff
@@ -3,12 +3,15 @@
...
 
 import requests
+import urllib3
 from celery import Task
 from django.conf import settings
 
 from posthog.celery import app
 from posthog.models import Action, Event, Team
 
+urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
```
Comment (macobo) [inline posthog/tasks/webhooks.py:13]: Q: Why is this needed?

File: posthog/models/action.py (hunk 3 above)
Comment (Twixes) [reply posthog/models/action.py:64]: Hm, they should still only get notifications for events since last calculation, since the filter below has that condition

Comment (macobo) [reply posthog/models/action.py:64]: If a full recalculation occurs, then above all rows first get deleted and then recreated. created_at will be greater than last_calculated_at for all of them.

File: posthog/tasks/webhooks.py (hunk 4 above)
Comment (Twixes) [reply posthog/tasks/webhooks.py:13]: It's a webhooks thing, we have `verify=false` on the request to Slack/Teams/Discord/whatever and `urllib3` complained in logs, but it's not a substantial issue and only was spamming stdout

**Review by macobo — COMMENTED** [review]
LGTM other than the unresolved question on webhooks. Once that's resolved, feel free to merge. :)

File: posthog/models/action.py (hunk 3 above)
Comment (macobo) [reply posthog/models/action.py:64]: Aha. That's fair if a bit tricky. Good chat, go get this merged!
